include/pisa/util/index_build_utils.hpp

```diff
@@ -161,4 +164,44 @@ void sample_inverted_index(const std::string &input_basename,
     }
 }
 
+
+template <typename SampleFn>
+void sample_inverted_index(std::string const &input_basename,
+                  std::string const &output_basename,
+                  SampleFn &&sample_fn)
+{
+
+    binary_freq_collection input(input_basename.c_str());
+
+    boost::filesystem::copy_file(fmt::format("{}.sizes", input_basename),
+                                 fmt::format("{}.sizes", output_basename),
+                                 boost::filesystem::copy_option::overwrite_if_exists);
+
+    std::ofstream dos(output_basename + ".docs");
+    std::ofstream fos(output_basename + ".freqs");
+
+    auto document_count = static_cast<uint32_t>(input.num_docs());
+    write_sequence(dos, gsl::make_span<uint32_t const>(&document_count, 1));
+
+    {
+        pisa::progress progress("Sampling inverted index", input.size());
+        for (auto const &plist : input) {
+            auto sample = sample_fn(plist.docs.size());
```

Comment:
I think in any reasonable case, we'll want the sampling to be stable, that is, we don't want to reorder the postings we keep. But here, we're opening ourselves to a situation where it's not ensured. I think it is reasonable to expect a stable sampling function but not to assume one is given. I suggest adding a comment at this function's signature that `sample_fn` must be stable, or that it must return sorted vectors, and here run `std::is_sorted` and fail if it's `false`. I think it's a reasonable trade-off (samples should be quite small) for the safety we get.

We could also just sort here if we want to be more flexible, but it's kind of wasteful, and I think it's reasonable to expect a sorted list of indices (the sampling method can sort itself if it knows it's not stable).